source/brailleDisplayDrivers/freedomScientific.py

```diff
@@ -1,292 +1,495 @@
+			if type(arg) == int:
+				return int2byte(arg)
+			return arg
+		arg1 = handleArg(arg1)
+		arg2 = handleArg(arg2)
+		arg3 = handleArg(arg3)
+		packet = [packetType, arg1, arg2, arg3, data]
+		if data:
+			packet.append(int2byte(self._calculateChecksum("".join(packet))))
+		self._dev.write("".join(packet))
+
+	def _onReceive(self, data):
+		if self.isUsb:
+			data = BytesIO(data)
+			packetType = data.read(1)
+		else:
+			packetType = data
+			data = self._dev
+
+		arg1 = data.read(1)
+		arg2 = data.read(1)
+		arg3 = data.read(1)
+		log.debug("Got packet of type %r with args: %r %r %r", packetType, arg1, arg2, arg3)
+		# Info and extended key responses are the only packets with payload and checksum
+		if packetType in (FS_PKT_INFO, FS_PKT_EXT_KEY):
+			length = ord(arg1)
+			payload = data.read(length)
+			checksum = ord(data.read(1))
+			calculatedChecksum = self._calculateChecksum(packetType + arg1 + arg2 + arg3 + payload)
+			assert calculatedChecksum == checksum, "Checksum mismatch, expected %s but got %s" % (checksum, payload[-1])
+		else:
+			payload = FS_DATA_EMPTY
+
+		self._handlePacket(packetType, arg1, arg2, arg3, payload)
+
+	def _handlePacket(self, packetType, arg1, arg2, arg3, payload):
+		"Handle a packet from the device"
+		if packetType == FS_PKT_ACK:
+			self._handleAck()
+		elif packetType == FS_PKT_NAK:
+			log.debugWarning("NAK received!")
+			self._handleAck()
+		elif packetType == FS_PKT_INFO:
+			self._manufacturer = payload[INFO_MANU_START:INFO_MANU_END].replace(FS_BYTE_NULL, "")
+			self._model = payload[INFO_MODEL_START:INFO_MODEL_END].replace(FS_BYTE_NULL, "")
+			self._firmwareVersion = payload[INFO_VERSION_START:INFO_VERSION_END].replace(FS_BYTE_NULL, "")
+			self.numCells = MODELS.get(self._model, 0)
+			if self.numCells in FOCUS_1_CELL_COUNTS:
+				# Focus 1: apply custom translation table
+				self.translationTable = FOCUS_1_TRANSLATION_TABLE
+			log.debug("Device info: manufacturer: %s model: %s, version: %s",
+				self._manufacturer, self._model, self._firmwareVersion)
+		elif packetType == FS_PKT_WHEEL:
+			# We use 0X7 to only get the last three bits
+			wheelNumber = ((ord(arg1) >> 3) & 0X7)
+			count = ord(arg1) & 0X7
+			if wheelNumber < 2:
+				isRight = False
+			else:
+				isRight = True
+			isDown = wheelNumber % 2 == 1
+			if isRight:
+				isDown = not isDown
+			for _i in xrange(count):
+				gesture = WizWheelGesture(self._model, isDown, isRight)
+				try:
+					inputCore.manager.executeGesture(gesture)
+				except inputCore.NoInputGestureAction:
+					pass
+		elif packetType == FS_PKT_BUTTON:
+			key = ord(arg1)
+			isPress = (ord(arg2) & 0X01) != 0
+			isTopRow = bool(ord(arg3))
+			if isPress:
+				# Ignore keypresses
+				return
+			gesture = RoutingGesture(self._model, key, isTopRow)
+			try:
+				inputCore.manager.executeGesture(gesture)
+			except inputCore.NoInputGestureAction:
+				pass
+		elif packetType == FS_PKT_KEY:
+			keyBits = ord(arg1) | (ord(arg2) << 8) | (ord(arg3) << 16)

```

Comment:
Do we know that each of these (arg1, arg2, arg3) will only hold an 8 bit number? I assume the intent is to construct one 24 bit number? And this allows a key press to be forwarded by the device? Do these match keyboard keycodes?

Reply:
I don't think they match PC keyboard keycodes. The 24-bit number encodes a combination of keys being pressed on the eight-key braille keyboard of the device.